reader.go

```diff
@@ -180,51 +187,68 @@ func (zr *Reader) Read(p []byte) (int, error) {
 		return 0, nil
 	}
 
-	if zr.outBuf.pos == zr.outBuf.size {
-		if err := zr.fillOutBuf(); err != nil {
+	if zr.readerPos >= len(zr.outBuf) {
+		if len(p) >= minDirectWriteBufferSize {
+			// write directly into the target buffer
+			// but make sure to override its capacity
+			return zr.fillOutBuf(p[:len(p):len(p)])
+		}
+		if _, err := zr.fillOutBuf(nil); err != nil {
 			return 0, err
 		}
+		zr.readerPos = 0
 	}
 
-	n := copy(p, zr.outBufGo[zr.outBuf.pos:zr.outBuf.size])
-	zr.outBuf.pos += C.size_t(n)
+	n := copy(p, zr.outBuf[zr.readerPos:])
+	zr.readerPos += n
 	return n, nil
 }
 
-func (zr *Reader) fillOutBuf() error {
-	if zr.inBuf.pos == zr.inBuf.size && zr.outBuf.size < dstreamOutBufSize {
+func (zr *Reader) fillOutBuf(target []byte) (int, error) {
+	dst := target
+	if dst == nil {
+		dst = zr.outBuf
+	}
+
+	if int(zr.sizes.srcPos) == len(zr.inBuf) && !zr.skipNextRead {
 		// inBuf is empty and the previously decompressed data size
-		// is smaller than the maximum possible zr.outBuf.size.
+		// is smaller than the maximum possible dst.size.
 		// This means that the internal buffer in zr.ds doesn't contain
 		// more data to decompress, so read new data into inBuf.
 		if err := zr.fillInBuf(); err != nil {
-			return err
+			return 0, err
 		}
 	}
 
+	zr.sizes.dstSize = C.size_t(cap(dst))
+	zr.sizes.dstPos = 0
+
+	inHdr := (*reflect.SliceHeader)(unsafe.Pointer(&zr.inBuf))
+	outHdr := (*reflect.SliceHeader)(unsafe.Pointer(&dst))
```

Comment:
you use the noescape() trick in compressInternal(), but not for dst here... I guess there is no chance stack allocated buffers could be used in this context?